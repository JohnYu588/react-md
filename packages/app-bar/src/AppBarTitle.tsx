import { Typography } from "@react-md/core";
import type { HTMLAttributes } from "react";
import { forwardRef } from "react";
import type { AppBarTitleClassNameOptions } from "./styles";
import { getAppBarTitleClassName } from "./styles";

export interface AppBarTitleProps
  extends HTMLAttributes<HTMLHeadingElement>,
    AppBarTitleClassNameOptions {}

export const AppBarTitle = forwardRef<HTMLHeadingElement, AppBarTitleProps>(
  function AppBarTitle(props, ref) {
    const {
      noWrap = false,
      keyline = "small",
      children,
      className,
      ...remaining
    } = props;
    return (
      <Typography
        ref={ref}
        {...remaining}
        type="headline-6"
        className={getAppBarTitleClassName({
          noWrap,
          keyline,
          className,
        })}
      >
        {children}
      </Typography>
    );
  }
);
