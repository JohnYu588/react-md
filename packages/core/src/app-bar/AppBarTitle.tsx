import { forwardRef } from "react";

import { type TextOverflow } from "../cssUtils.js";
import { Typography, type TypographyProps } from "../typography/Typography.js";
import { type TypographyType } from "../typography/typographyStyles.js";
import { type AppBarTitleClassNameOptions, appBarTitle } from "./styles.js";

/**
 * @since 6.0.0 The `keyline` prop was changed from a boolean to a type
 * union of different keylines: {@link AppBarTitleKeyline}
 * @since 6.0.0 The `noWrap` prop was removed in favor of the `textOverflow`
 * prop inherited through the base `Typography` component. The `textOverflow`
 * will default to `"ellipsis"` which is new as well.
 */
export interface AppBarTitleProps
  extends TypographyProps,
    AppBarTitleClassNameOptions {
  /** @defaultValue `"headline-6"` */
  type?: TypographyType;

  /**
   * @defaultValue `"ellipsis"`
   */
  textOverflow?: TextOverflow;
}

/**
 * @example Updating the Keyline
 * ```tsx
 * import { AppBar } from "@react-md/core/app-bar/AppBar";
 * import { AppBarTitle } from "@react-md/core/app-bar/AppBarTitle";
 * import type { ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   return (
 *     <AppBar>
 *       <AppBarTitle keyline="nav">
 *         Offset as if there was a nav button to the left
 *       </AppBarTitle>
 *     </AppBar>
 *   );
 * }
 * ```
 *
 * @since 6.0.0 The `keyline` prop was changed from a boolean to a type
 * union of different keylines: {@link AppBarTitleKeyline}
 * @since 6.0.0 The `noWrap` prop was removed in favor of the `textOverflow`
 * prop inherited through the base `Typography` component. The `textOverflow`
 * will default to `"ellipsis"` which is new as well.
 */
export const AppBarTitle = forwardRef<HTMLHeadingElement, AppBarTitleProps>(
  function AppBarTitle(props, ref) {
    const {
      type = "headline-6",
      keyline = "small",
      children,
      className,
      textOverflow = "ellipsis",
      ...remaining
    } = props;
    return (
      <Typography
        {...remaining}
        ref={ref}
        type={type}
        className={appBarTitle({
          keyline,
          className,
        })}
        textOverflow={textOverflow}
      >
        {children}
      </Typography>
    );
  }
);
