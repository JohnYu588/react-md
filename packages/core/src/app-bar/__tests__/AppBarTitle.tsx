import { describe, expect, it } from "@jest/globals";
import { createRef } from "react";
import { render } from "../../test-utils/index.js";

import { appBarTitle, AppBarTitle } from "../AppBarTitle.js";

describe("AppBarTitle", () => {
  it("should apply the correct styling, HTML attributes, and allow a ref", () => {
    const ref = createRef<HTMLHeadingElement>();
    const { getByRole, rerender } = render(
      <AppBarTitle ref={ref}>Title</AppBarTitle>
    );

    const title = getByRole("heading", { name: "Title" });
    expect(ref.current).toBeInstanceOf(HTMLHeadingElement);
    expect(ref.current).toBe(title);
    expect(title).toMatchSnapshot();

    rerender(
      <AppBarTitle style={{ color: "aqua" }} className="custom-class-name">
        Title
      </AppBarTitle>
    );
    expect(title).toMatchSnapshot();

    rerender(<AppBarTitle keyline="nav">Title</AppBarTitle>);
    expect(title).toMatchSnapshot();

    rerender(<AppBarTitle keyline="list">Title</AppBarTitle>);
    expect(title).toMatchSnapshot();

    rerender(<AppBarTitle lineWrap>Title</AppBarTitle>);
    expect(title).toMatchSnapshot();
  });

  describe("styling utils", () => {
    it("should be callable without any arguments", () => {
      expect(appBarTitle()).toMatchSnapshot();
    });
  });
});
