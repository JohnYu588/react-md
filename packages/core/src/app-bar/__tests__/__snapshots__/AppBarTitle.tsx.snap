// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppBarTitle should apply the correct styling, HTML attributes, and allow a ref 1`] = `
<h6
  class="rmd-typography rmd-typography--headline-6 rmd-nowrap rmd-ellipsis rmd-app-bar-title"
>
  Title
</h6>
`;

exports[`AppBarTitle should apply the correct styling, HTML attributes, and allow a ref 2`] = `
<h6
  class="rmd-typography rmd-typography--headline-6 rmd-nowrap rmd-ellipsis rmd-app-bar-title custom-class-name"
  style="color: aqua;"
>
  Title
</h6>
`;

exports[`AppBarTitle should apply the correct styling, HTML attributes, and allow a ref 3`] = `
<h6
  class="rmd-typography rmd-typography--headline-6 rmd-nowrap rmd-ellipsis rmd-app-bar-title rmd-app-bar-title--nav-keyline"
  style=""
>
  Title
</h6>
`;

exports[`AppBarTitle should apply the correct styling, HTML attributes, and allow a ref 4`] = `
<h6
  class="rmd-typography rmd-typography--headline-6 rmd-nowrap rmd-ellipsis rmd-app-bar-title rmd-app-bar-title--keyline"
  style=""
>
  Title
</h6>
`;

exports[`AppBarTitle should apply the correct styling, HTML attributes, and allow a ref 5`] = `
<h6
  class="rmd-typography rmd-typography--headline-6 rmd-app-bar-title"
  style=""
>
  Title
</h6>
`;

exports[`AppBarTitle styling utils should be callable without any arguments 1`] = `"rmd-app-bar-title"`;
